### PR TITLE
Add upload script for mzML files

### PR DESCRIPTION
**Pull request by blink1073**
Provides a config and script for loading mzML files to the workspace.

### CODE REVIEW

**Comment by blink1073** [comment]
cc @benbowen

**Comment by msneddon** [comment]
@jkbaumohl This is the PR I mentioned to you last week.  It would be great if we can get your help to review the changes and hopefully test things out locally.  Thanks!

Note that this PR is related to: https://atlassian.kbase.us/browse/KBASE-1853

File: plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            "shock_id",
+            "handle_id",
+            "output_file_name",
+            "input_mapping",
+            "name",
+            "polarity",
+            "group",
+            "inclusion_order",
+            "retention_correction",
+            "normalization_factor",
```
Comment (rsutormin) [inline plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json]: Comma should be removed at the end of this line.

File: plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            "type": "string",
+            "required": false,
+            "help": "Run retention correction."
+        },
+        {
+            "name": "normalization_factor",
+            "type": "string",
+            "required": false,
+            "help": "Run normalization factor"
+        },
```
Comment (rsutormin) [inline plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json]: Comma should be removed at the end of this line.

File: plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+import simplejson
+
+# KBase imports
+try:
+    import biokbase.Transform.script_utils as script_utils
+except ImportError:
+    from . import script_utils
+
+# Metatlas Imports
+from metatlas import mzml_loader
```
Comment (rsutormin) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: This dependency is not found in KBase environment. You should add something like this into https://github.com/kbase/transform/blob/develop/deps/pylib.sh :
-----------------------------------------------------------------------------------------
pip install git+https://github.com/metabolite-atlas/metatlas
But this command doesn't work unfortunately with such error:
Downloading/unpacking git+https://github.com/metabolite-atlas/metatlas
  Cloning https://github.com/metabolite-atlas/metatlas to /tmp/pip-PAM0Ze-build
  Running setup.py (path:/tmp/pip-PAM0Ze-build/setup.py) egg_info for package from git+https://github.com/metabolite-atlas/metatlas
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ImportError: No module named pymzml
    cp: missing destination file operand after `obo.patch'
    Try `cp --help' for more information.
    install.sh: line 9: pushd: no other directory
    install.sh: line 10: obo.patch: No such file or directory
    rm: cannot remove `obo.patch': No such file or directory
    install.sh: line 12: popd: directory stack empty
Downloading/unpacking pymzml (from metatlas==0.2)
  Downloading pymzml-0.7.5.zip (115kB): 115kB downloaded
  Running setup.py (path:/tmp/pip_build_root/pymzml/setup.py) egg_info for package pymzml
Requirement already satisfied (use --upgrade to upgrade): simplejson in <...> (from metatlas==0.2)
Installing collected packages: pymzml, metatlas
  Running setup.py install for pymzml
  Running setup.py install for metatlas
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ImportError: No module named pymzml
    cp: missing destination file operand after `obo.patch'
    Try `cp --help' for more information.
    install.sh: line 9: pushd: no other directory
    install.sh: line 10: obo.patch: No such file or directory
    rm: cannot remove `obo.patch': No such file or directory
    install.sh: line 12: popd: directory stack empty

File: lib/biokbase/Transform/script_utils.py [hunk 4]
```diff
@@ -136,7 +136,7 @@ def parse_docs(docstring=None):
     remainder = script_details["Args"]
     argument_values = list()
     for k in reversed(argument_keys):
-        remainder, value = remainder.split(k)        
+        remainder, _, value = remainder.rpartition(k)
```
Comment (rsutormin) [inline lib/biokbase/Transform/script_utils.py]: Is this feature absolutely necessary for rest of the code? If not I would try to keep this change in separate PR. It will be better in a way you don't create unnecessary bindings and you let the owner of repo to decide independently.

**Comment by msneddon** [comment]
@blink1073 @benbowen I guess it's probably obvious from these comments, but Roman is also going to try and help us accelerate things with the metatlas integration.

File: plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json [hunk 5]
```diff
@@ -0,0 +1,84 @@
+{
```
Comment (rsutormin) [inline plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json]: Is it possible to rename this file from
    mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json
into
    mzML.LCMS_to_MetaboliteAtlas2.MAFileInfo.json
?
It's not so much required. It's just keeping naming system consistent (take a look at other config file names).

Comment (jkbaumohl) [inline plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json]: Matt has final say.  But if it is more consistent naming I would think he
would be for it.

Comment (rsutormin) [inline plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json]: @mlhenderson Matt, what's your reference in this case?

File: lib/biokbase/Transform/script_utils.py (hunk 4 above)
Comment (blink1073) [inline lib/biokbase/Transform/script_utils.py]: This is needed because one of our fields is named `name`, which is what caused the conflict leading to the other PR.

File: plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py (hunk 3 above)
Comment (blink1073) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: Should be fixed now.

File: deps/pylib.sh [hunk 6]
```diff
@@ -1 +1,2 @@
 pip install requests_toolbelt filemagic ftputil
+pip install git+https://github.com/metabolite-atlas/metatlas
```
Comment (rsutormin) [inline deps/pylib.sh]: Ok, now pip installation finishes successfully. But "metatlas" folder doesn't appear in python site-packages. There is only "metatlas-0.2-py2.7.egg-info" folder with dependencies (and other text files). So in particular there is no "mzml_loader.py" anywhere. I tried it in two different architectures (linux and mac). Did you try to make this command on clean python environment?

Comment (blink1073) [inline deps/pylib.sh]: Fixed, I tried it from a clean python environment.

Comment (rsutormin) [inline deps/pylib.sh]: Ok, next error is:
```
Traceback (most recent call last):
  File "/kb/deployment/pybin/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py", line 19, in <module>
    from metatlas import mzml_loader
  File "/kb/runtime/lib/python2.7/site-packages/metatlas/__init__.py", line 2, in <module>
    from .mzml_loader import mzml_to_hdf
  File "/kb/runtime/lib/python2.7/site-packages/metatlas/mzml_loader.py", line 9, in <module>
    import tables
ImportError: No module named tables
```
I think ```tables``` should be added into first line ```pip install ...```. But it's not enough. Because when I just run ```pip install tables``` I get such error:
```
Downloading/unpacking tables
  Downloading tables-3.2.0.tar.gz (7.0MB): 7.0MB downloaded
  Running setup.py (path:/tmp/pip_build_root/tables/setup.py) egg_info for package tables
    /usr/bin/ld: cannot find -lhdf5
    collect2: ld returned 1 exit status
    * Using Python 2.7.3 (default, Jan  9 2014, 18:36:02)
    * USE_PKGCONFIG: False
    .. ERROR:: Could not find a local HDF5 installation.
       You may need to explicitly state where your local HDF5 headers and
       library can be found by setting the ``HDF5_DIR`` environment
       variable or by using the ``--hdf5`` command-line option.
    Complete output from command python setup.py egg_info:
    /usr/bin/ld: cannot find -lhdf5
collect2: ld returned 1 exit status
* Using Python 2.7.3 (default, Jan  9 2014, 18:36:02)
* USE_PKGCONFIG: False
.. ERROR:: Could not find a local HDF5 installation.
   You may need to explicitly state where your local HDF5 headers and
   library can be found by setting the ``HDF5_DIR`` environment
   variable or by using the ``--hdf5`` command-line option.
```

Comment (blink1073) [inline deps/pylib.sh]: Tables is a tricky one.  It is already installed in the narrative environment.  Are you running this code on dev02?

Comment (rsutormin) [inline deps/pylib.sh]: It doesn't matter. We need to have some temporary instructions until hdf5 is added into kbase runtime (which could take long time actually). It means there should be script in "transform/deps" folder installing it manually. I'll help you with it. What version of hdf5 do you prefer to use? If this script doesn't require sudo privileges then you just run it from make deploy target. If it does (like apt-get) then you mention about this script in readme file as preinstalled requirement. It's better to do it ones and make sure other people won't struggle with it hating you.

Comment (blink1073) [inline deps/pylib.sh]: What I meant it that HDF is already installed in the kbase runtime. I can `import tables` from a narrative.  In your case `apt-get` is required, it shouldn't matter which `libhdf`. 



—

Comment (blink1073) [inline deps/pylib.sh]: Or are the runtime and the narrative distinct environments?



—

**Comment by rsutormin** [comment]
I added script with installation commands for 'tables' lib in this PR: https://github.com/metabolite-atlas/transform/pull/1

File: plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                        action='store', type=int, required=False)
+    parser.add_argument('--group',
+                        help=script_details["Args"]["group"],
+                        action='store', type=str, required=False)
+    parser.add_argument('--inclusion_order',
+                        help=script_details["Args"]["inclusion_order"],
+                        action='store', type=int, required=False)
+    parser.add_argument('--retention_correction',
+                        help=script_details["Args"]["retention_correction"],
+                        action='store', type=float, equired=False)
```
Comment (rsutormin) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: "equired" -> "required"

File: plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    parser.add_argument('--atlases',
+                        help=script_details["Args"]["atlases"],
+                        action='store', type=str, nargs='?',
+                        required=False)
+    parser.add_argument('--name',
+                        help=script_details["Args"]["name"],
+                        action='store', type=str, equired=False)
```
Comment (rsutormin) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: "equired" -> "required"

Comment (blink1073) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: Fixed.

File: plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            run_info['group'] = group
+        if inclusion_order is not None:
+            run_info['inclusion_order'] = inclusion_order
+        if normalization_factor is not None:
+            run_info['normalization_factor'] = normalization_factor
+        if retention_correction is not None:
+            run_info['retention_correction'] = retention_correction
+
+        if shock_service_url:
+            run_info["run_file_id"] = shock_info["id"]
```
Comment (rsutormin) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: And one more fix. Instead of putting shock node id into handle field you should create handle pointing to that shock node and use id of this handle in workspace object. Can you substitute this line:
```
            run_info["run_file_id"] = shock_info["id"]
```
by following lines:
```
            handle_id = script_utils.getHandles(logger, shock_service_url,
                    handle_service_url, [shock_info["id"]], token=token)[0]
            run_info["run_file_id"] = handle_id
```
?

Comment (blink1073) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: Done.

File: deps/sudo_install_tables.sh [hunk 10]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
Comment (rsutormin) [inline deps/sudo_install_tables.sh]: One more correction. Can we add these lines before any installation lines to prevent unnecessary sudo calls if tables lib is already installed?
```
python_tables_installed=`pip list | grep '^tables (' | wc -l`
if [ "$python_tables_installed" > "0" ]; then
    exit 0
fi
echo "Installing python tables"
```

File: deps/pylib.sh (hunk 6 above)
Comment (rsutormin) [inline deps/pylib.sh]: And let's check if metatlas was already installed. So instead of
```
pip install git+https://github.com/metabolite-atlas/metatlas
```
we can put this:
```
python_metatlas_installed=`pip list | grep '^metatlas (' | wc -l`
if [ "$python_metatlas_installed" == "0" ]; then
    pip install git+https://github.com/metabolite-atlas/metatlas
fi
```

File: deps/sudo_install_tables.sh (hunk 10 above)
Comment (blink1073) [inline deps/sudo_install_tables.sh]: All set.

File: lib/biokbase/Transform/script_utils.py (hunk 4 above)
Comment (rsutormin) [inline lib/biokbase/Transform/script_utils.py]: Is it possible to rename ```name``` field into something like ```mzml_file_name```? It will allow us to be independent in this PR from changes in core file like ```script_utils.py``` which is not so easy to get merging approval for.

Comment (mlhenderson) [inline lib/biokbase/Transform/script_utils.py]: I have a fix for the problem this is meant to address which I will submit a
PR for review.

**Comment by blink1073** [comment]
I reverted the change to `script_utils.py` (other than a trailing whitespace), changed the name of the attribute, and updated the spec.

File: plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    try:
+        transform(shock_service_url=args.shock_service_url,
+                  handle_service_url=args.handle_service_url,
+                  output_file_name=args.output_file_name,
+                  input_directory=args.input_directory,
+                  working_directory=args.working_directory,
+                  shock_id=args.shock_id,
+                  handle_id=args.handle_id,
+                  input_mapping=args.input_mapping,
+                  mzml_file_name=args.mzml_files,
```
Comment (rsutormin) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: Next error is:
```
Traceback (most recent call last):
  File "/kb/deployment/pybin/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py", line 222, in main
    mzml_file_name=args.mzml_files,
AttributeError: 'Namespace' object has no attribute 'mzml_files'
```

File: plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+            "working_directory"
+        ],
+        "optional_fields":
+        [
+            "handle_service_url",
+            "shock_id",
+            "handle_id",
+            "output_file_name",
+            "input_mapping",
+            "name",
```
Comment (rsutormin) [inline plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json]: Shouldn't it be ```"mzml_file_name",```?

File: plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    "user_options":
+    [
+        {
+            "name": "output_file_name",
+            "type": "string",
+            "required": true,
+            "help": "If not supplied, will default the base input_file_name with '_MAFileInfo.json' appended to the end."
+        },
+        {
+            "name": "name",
```
Comment (rsutormin) [inline plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json]: Shouldn't it be ```"name": "mzml_file_name"```?

Comment (blink1073) [inline plugins/configs/mzML.LCMS_to_MetaboliteAtlas2_MAFileInfo.json]: Fixed.

File: plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+                            .format(path))
+
+        hdf_file = mzml_loader.mzml_to_hdf(path)
+
+        if shock_service_url:
+            shock_info = script_utils.upload_file_to_shock(logger,
+                    shock_service_url, hdf_file, token=token)
+
+        run_info = dict()
+        run_info['name'] = name or fname.replace('.mzML', '')
```
Comment (rsutormin) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: And one more lost ```name``` ref: https://github.com/kbase/transform/pull/151/files#diff-a6e65713be5cb812af79d41b4d89fef5R112

Comment (rsutormin) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: And ```'name'``` key is now changed here: https://narrative-ci.kbase.us/functional-site/#/spec/type/MetaboliteAtlas2.MAFileInfo

Comment (blink1073) [inline plugins/scripts/upload/trns_transform_mzML_LCMS_to_MetaboliteAtlas2_MAFileInfo.py]: Fixed.

**Comment by rsutormin** [comment]
Ok, now this PR looks good and ready to be merged. Matt, do you have any suggestions? @mlhenderson
